### ISSUE TITLE
Recently merged canary branch has no tests

### ISSUE THREAD

**Issue by nosmo** [equalitie/edgemanage#20]
The new functionality should have tests. Perhaps after a refactor as per #18

**Comment by graphiclunarkid** [comment]
I agree good test-coverage is important, however I think refactoring without tests would be more difficult and risky, so I suggest writing the tests first.

1. Achieve good test coverage for the current code-base.
2. Make sure the tests pass (i.e. fix any minor bugs revealed by the new tests).
3. Refactor and break stuff. Continue refactoring until the new code passes all the tests (#18).

This approach uses the tests as a measure of whether the refactoring has been done correctly. If the tests pass before the refactoring starts, and they pass again with the new code, we can be reasonably sure the refactored code is behaving the same way as the old code did.
